File: basic_dsa_twenty_three.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if a Number is Spy Number
# A Spy number is a number where the sum of its digits equals the product of its digits
# Example: 1124 = 1+1+2+4 = 8 and 1*1*2*4 = 8, so 1124 is a Spy number

def sum_of_digits(n):
    """
    Calculate sum of digits of a number
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    total = 0
    num = abs(n)
    
    while num > 0:
        total += num % 10
        num //= 10
    
    return total


def product_of_digits(n):
    """
    Calculate product of digits of a number
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    product = 1
    num = abs(n)
    
    # Handle case where number contains 0
    while num > 0:
        digit = num % 10
        product *= digit
        num //= 10
    
    return product


def is_spy_number(n):
    """
    Check if a number is Spy number
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return False
    
    return sum_of_digits(n) == product_of_digits(n)


def get_spy_numbers_in_range(start, end):
    """
    Find all Spy numbers in a given range
    Time Complexity: O((end - start) * log n)
    Space Complexity: O(k) where k is the number of Spy numbers
    """
    spy_numbers = []
    
    for num in range(start, end + 1):
        if is_spy_number(num):
            spy_numbers.append(num)
    
    return spy_numbers


# Problem 2: Convert Hexadecimal to Decimal
# Convert a hexadecimal (base 16) number to its decimal representation

def hexadecimal_to_decimal_iterative(hex_str):
    """
    Convert hexadecimal string to decimal using iteration
    Time Complexity: O(n) where n is length of hex string
    Space Complexity: O(1)
    """
    if not hex_str or hex_str == "0":
        return 0
    
    hex_digits = "0123456789ABCDEF"
    is_negative = hex_str[0] == '-'
    start = 1 if is_negative else 0
    
    decimal = 0
    power = 0
    
    # Process from right to left
    for i in range(len(hex_str) - 1, start - 1, -1):
        char = hex_str[i].upper()
        if char in hex_digits:
            digit_value = hex_digits.index(char)
            decimal += digit_value * (16 ** power)
            power += 1
        else:
            raise ValueError(f"Invalid hexadecimal digit: {char}")
    
    return -decimal if is_negative else decimal


def hexadecimal_to_decimal_recursive(hex_str):
    """
    Convert hexadecimal string to decimal using recursion
    Time Complexity: O(n) where n is length of hex string
    Space Complexity: O(n) due to recursion stack
    """
    hex_digits = "0123456789ABCDEF"
    
    def helper(hex_str, index, power):
        if index < 0 or (index == 0 and hex_str[0] == '-'):
            return 0
        char = hex_str[index].upper()
        if char not in hex_digits:
            raise ValueError(f"Invalid hexadecimal digit: {char}")
        digit_value = hex_digits.index(char)
        return digit_value * (16 ** power) + helper(hex_str, index - 1, power + 1)
    
    if not hex_str or hex_str == "0":
        return 0
    
    is_negative = hex_str[0] == '-'
    start = len(hex_str) - 1
    
    result = helper(hex_str, start, 0)
    return -result if is_negative else result


def hexadecimal_to_decimal_builtin(hex_str):
    """
    Convert hexadecimal string to decimal using built-in function
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not hex_str:
        return 0
    
    is_negative = hex_str[0] == '-'
    hex = hex_str[1:] if is_negative else hex_str
    
    decimal = int(hex, 16)
    return -decimal if is_negative else decimal


def hexadecimal_array_to_decimal(hex_array):
    """
    Convert hexadecimal array (list of characters) to decimal
    Time Complexity: O(n) where n is length of array
    Space Complexity: O(1)
    """
    if not hex_array:
        return 0
    
    hex_digits = "0123456789ABCDEF"
    decimal = 0
    power = len(hex_array) - 1
    
    for char in hex_array:
        char_upper = char.upper()
        if char_upper in hex_digits:
            digit_value = hex_digits.index(char_upper)
            decimal += digit_value * (16 ** power)
            power -= 1
        else:
            raise ValueError(f"Invalid hexadecimal digit: {char}")
    
    return decimal


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if a Number is Spy Number")
    print("=" * 60)
    
    # Test Case 1 - Spy number 1124
    num1 = 1124
    result1 = is_spy_number(num1)
    sum1 = sum_of_digits(num1)
    product1 = product_of_digits(num1)
    print(f"Number: {num1}")
    print(f"Sum of digits: {sum1}")
    print(f"Product of digits: {product1}")
    print(f"Is Spy number: {result1}")
    
    # Test Case 2 - Spy number 22
    num2 = 22
    result2 = is_spy_number(num2)
    sum2 = sum_of_digits(num2)
    product2 = product_of_digits(num2)
    print(f"\nNumber: {num2}")
    print(f"Sum of digits: {sum2}")
    print(f"Product of digits: {product2}")
    print(f"Is Spy number: {result2}")
    
    # Test Case 3 - Not Spy number
    num3 = 123
    result3 = is_spy_number(num3)
    sum3 = sum_of_digits(num3)
    product3 = product_of_digits(num3)
    print(f"\nNumber: {num3}")
    print(f"Sum of digits: {sum3}")
    print(f"Product of digits: {product3}")
    print(f"Is Spy number: {result3}")
    
    # Test Case 4 - Single digit (always Spy)
    num4 = 1
    result4 = is_spy_number(num4)
    print(f"\nNumber: {num4}")
    print(f"Is Spy number: {result4}")
    
    # Test Case 5 - Find Spy numbers in range
    start = 1
    end = 2000
    spy_nums = get_spy_numbers_in_range(start, end)
    print(f"\nSpy numbers in range [{start}, {end}]:")
    print(spy_nums[:20])  # Print first 20
    print(f"Total count: {len(spy_nums)}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Convert Hexadecimal to Decimal")
    print("=" * 60)
    
    # Test Case 1 - Basic hexadecimal
    hex1 = "FF"
    result1_iter = hexadecimal_to_decimal_iterative(hex1)
    result1_rec = hexadecimal_to_decimal_recursive(hex1)
    result1_builtin = hexadecimal_to_decimal_builtin(hex1)
    print(f"Hexadecimal: {hex1}")
    print(f"Decimal (iterative): {result1_iter}")
    print(f"Decimal (recursive): {result1_rec}")
    print(f"Decimal (built-in): {result1_builtin}")
    
    # Test Case 2 - Zero
    hex2 = "0"
    result2 = hexadecimal_to_decimal_iterative(hex2)
    print(f"\nHexadecimal: {hex2}")
    print(f"Decimal: {result2}")
    
    # Test Case 3 - Large hexadecimal
    hex3 = "ABCD"
    result3 = hexadecimal_to_decimal_iterative(hex3)
    print(f"\nHexadecimal: {hex3}")
    print(f"Decimal: {result3}")
    
    # Test Case 4 - Hexadecimal array
    hex4 = ['1', 'A', 'F']
    result4 = hexadecimal_array_to_decimal(hex4)
    print(f"\nHexadecimal (array): {hex4}")
    print(f"Decimal: {result4}")
    
    # Test Case 5 - Lowercase hexadecimal
    hex5 = "abc"
    result5 = hexadecimal_to_decimal_iterative(hex5)
    print(f"\nHexadecimal: {hex5}")
    print(f"Decimal: {result5}")

File: basic_dsa_twenty_three.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_twenty_three {
    
    // Problem 1: Check if a Number is Spy Number
    // A Spy number is a number where the sum of its digits equals the product of its digits
    // Example: 1124 = 1+1+2+4 = 8 and 1*1*2*4 = 8, so 1124 is a Spy number
    
    /**
     * Calculate sum of digits of a number
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static int sumOfDigits(int n) {
        int total = 0;
        int num = Math.abs(n);
        
        while (num > 0) {
            total += num % 10;
            num /= 10;
        }
        
        return total;
    }
    
    /**
     * Calculate product of digits of a number
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static int productOfDigits(int n) {
        int product = 1;
        int num = Math.abs(n);
        
        while (num > 0) {
            int digit = num % 10;
            product *= digit;
            num /= 10;
        }
        
        return product;
    }
    
    /**
     * Check if a number is Spy number
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static boolean isSpyNumber(int n) {
        if (n <= 0) {
            return false;
        }
        
        return sumOfDigits(n) == productOfDigits(n);
    }
    
    /**
     * Find all Spy numbers in a given range
     * Time Complexity: O((end - start) * log n)
     * Space Complexity: O(k) where k is the number of Spy numbers
     */
    public static ArrayList<Integer> getSpyNumbersInRange(int start, int end) {
        ArrayList<Integer> spyNumbers = new ArrayList<>();
        
        for (int num = start; num <= end; num++) {
            if (isSpyNumber(num)) {
                spyNumbers.add(num);
            }
        }
        
        return spyNumbers;
    }
    
    // Problem 2: Convert Hexadecimal to Decimal
    // Convert a hexadecimal (base 16) number to its decimal representation
    
    /**
     * Convert hexadecimal string to decimal using iteration
     * Time Complexity: O(n) where n is length of hex string
     * Space Complexity: O(1)
     */
    public static int hexadecimalToDecimalIterative(String hexStr) {
        if (hexStr == null || hexStr.isEmpty() || hexStr.equals("0")) {
            return 0;
        }
        
        String hexDigits = "0123456789ABCDEF";
        boolean isNegative = hexStr.charAt(0) == '-';
        int start = isNegative ? 1 : 0;
        
        int decimal = 0;
        int power = 0;
        
        // Process from right to left
        for (int i = hexStr.length() - 1; i >= start; i--) {
            char ch = Character.toUpperCase(hexStr.charAt(i));
            int digitValue = hexDigits.indexOf(ch);
            if (digitValue == -1) {
                throw new IllegalArgumentException("Invalid hexadecimal digit: " + ch);
            }
            decimal += digitValue * Math.pow(16, power);
            power++;
        }
        
        return isNegative ? -decimal : decimal;
    }
    
    /**
     * Convert hexadecimal string to decimal using recursion
     * Time Complexity: O(n) where n is length of hex string
     * Space Complexity: O(n) due to recursion stack
     */
    public static int hexadecimalToDecimalRecursive(String hexStr) {
        if (hexStr == null || hexStr.isEmpty() || hexStr.equals("0")) {
            return 0;
        }
        
        boolean isNegative = hexStr.charAt(0) == '-';
        String hex = isNegative ? hexStr.substring(1) : hexStr;
        
        int result = helper(hex, hex.length() - 1, 0);
        return isNegative ? -result : result;
    }
    
    private static int helper(String hex, int index, int power) {
        String hexDigits = "0123456789ABCDEF";
        if (index < 0) {
            return 0;
        }
        char ch = Character.toUpperCase(hex.charAt(index));
        int digitValue = hexDigits.indexOf(ch);
        if (digitValue == -1) {
            throw new IllegalArgumentException("Invalid hexadecimal digit: " + ch);
        }
        return digitValue * (int) Math.pow(16, power) + helper(hex, index - 1, power + 1);
    }
    
    /**
     * Convert hexadecimal string to decimal using built-in function
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static int hexadecimalToDecimalBuiltin(String hexStr) {
        if (hexStr == null || hexStr.isEmpty()) {
            return 0;
        }
        
        boolean isNegative = hexStr.charAt(0) == '-';
        String hex = isNegative ? hexStr.substring(1) : hexStr;
        
        int decimal = Integer.parseInt(hex, 16);
        return isNegative ? -decimal : decimal;
    }
    
    /**
     * Convert hexadecimal array (list of characters) to decimal
     * Time Complexity: O(n) where n is length of array
     * Space Complexity: O(1)
     */
    public static int hexadecimalArrayToDecimal(char[] hexArray) {
        if (hexArray == null || hexArray.length == 0) {
            return 0;
        }
        
        String hexDigits = "0123456789ABCDEF";
        int decimal = 0;
        int power = hexArray.length - 1;
        
        for (char ch : hexArray) {
            char chUpper = Character.toUpperCase(ch);
            int digitValue = hexDigits.indexOf(chUpper);
            if (digitValue == -1) {
                throw new IllegalArgumentException("Invalid hexadecimal digit: " + ch);
            }
            decimal += digitValue * Math.pow(16, power);
            power--;
        }
        
        return decimal;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Check if a Number is Spy Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Spy number 1124
        int num1 = 1124;
        boolean result1 = isSpyNumber(num1);
        int sum1 = sumOfDigits(num1);
        int product1 = productOfDigits(num1);
        System.out.println("Number: " + num1);
        System.out.println("Sum of digits: " + sum1);
        System.out.println("Product of digits: " + product1);
        System.out.println("Is Spy number: " + result1);
        
        // Test Case 2 - Spy number 22
        int num2 = 22;
        boolean result2 = isSpyNumber(num2);
        int sum2 = sumOfDigits(num2);
        int product2 = productOfDigits(num2);
        System.out.println("\nNumber: " + num2);
        System.out.println("Sum of digits: " + sum2);
        System.out.println("Product of digits: " + product2);
        System.out.println("Is Spy number: " + result2);
        
        // Test Case 3 - Not Spy number
        int num3 = 123;
        boolean result3 = isSpyNumber(num3);
        int sum3 = sumOfDigits(num3);
        int product3 = productOfDigits(num3);
        System.out.println("\nNumber: " + num3);
        System.out.println("Sum of digits: " + sum3);
        System.out.println("Product of digits: " + product3);
        System.out.println("Is Spy number: " + result3);
        
        // Test Case 4 - Single digit (always Spy)
        int num4 = 1;
        boolean result4 = isSpyNumber(num4);
        System.out.println("\nNumber: " + num4);
        System.out.println("Is Spy number: " + result4);
        
        // Test Case 5 - Find Spy numbers in range
        int start = 1;
        int end = 2000;
        ArrayList<Integer> spyNums = getSpyNumbersInRange(start, end);
        System.out.println("\nSpy numbers in range [" + start + ", " + end + "] (first 20):");
        for (int i = 0; i < Math.min(20, spyNums.size()); i++) {
            System.out.print(spyNums.get(i) + " ");
        }
        System.out.println("\nTotal count: " + spyNums.size());
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Convert Hexadecimal to Decimal");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Basic hexadecimal
        String hex1 = "FF";
        int result1_dec = hexadecimalToDecimalIterative(hex1);
        int result1_rec = hexadecimalToDecimalRecursive(hex1);
        int result1_builtin = hexadecimalToDecimalBuiltin(hex1);
        System.out.println("Hexadecimal: " + hex1);
        System.out.println("Decimal (iterative): " + result1_dec);
        System.out.println("Decimal (recursive): " + result1_rec);
        System.out.println("Decimal (built-in): " + result1_builtin);
        
        // Test Case 2 - Zero
        String hex2 = "0";
        int result2_dec = hexadecimalToDecimalIterative(hex2);
        System.out.println("\nHexadecimal: " + hex2);
        System.out.println("Decimal: " + result2_dec);
        
        // Test Case 3 - Large hexadecimal
        String hex3 = "ABCD";
        int result3_dec = hexadecimalToDecimalIterative(hex3);
        System.out.println("\nHexadecimal: " + hex3);
        System.out.println("Decimal: " + result3_dec);
        
        // Test Case 4 - Hexadecimal array
        char[] hex4 = {'1', 'A', 'F'};
        int result4_dec = hexadecimalArrayToDecimal(hex4);
        System.out.println("\nHexadecimal (array): " + new String(hex4));
        System.out.println("Decimal: " + result4_dec);
        
        // Test Case 5 - Lowercase hexadecimal
        String hex5 = "abc";
        int result5_dec = hexadecimalToDecimalIterative(hex5);
        System.out.println("\nHexadecimal: " + hex5);
        System.out.println("Decimal: " + result5_dec);
    }
}
